src/node_crypto_common.cc

```diff
@@ -968,6 +968,16 @@ MaybeLocal<Value> GetPeerCert(
   return result;
 }
 
+MaybeLocal<Value> X509ToPEM(Environment* env, X509* cert) {
+  BIOPointer bio(BIO_new(BIO_s_mem()));
+
+  if (PEM_write_bio_X509(bio.get(), cert) == 0) {
+    return Undefined(env->isolate());
```

Comment:
This seems questionable. Why not return `MaybeLocal<Value>()`?

Reply:
… because there’s no JS exception pending in that case?

Reply:
Right, so it should probably call `ThrowCryptoError(env, ERR_get_error(), "PEM_write_bio_X509")` first.

Reply:
I originally wanted to use `ThrowCryptoError` there. The CPP_STYLE_GUIDE mentions that throwing exceptions from nested C++ methods should be avoided. I wasn't entirely sure whether I should absorb the true error (current PR), try to return the error out via a parameter, throw from the nested function, or get rid of the nested function entirely. `ThrowCryptoError` isn't callable from `node_crypto_common.cc` (yet) - it's still in `node_crypto.cc`

The refactored/new `node_crypto_common.cc` seems to (mostly) avoid nested throws. `X509ToObject` doesn't check error conditions for most of its OpenSSL calls. 

I feel that the most correct code would probably be either to throw from `node_crypto_common` and establish that as a valid pattern for truly exceptional conditions or just return `MaybeLocal<Value>()` and have the original OpenSSL error condition "lost", but not ignored. 

Unfortunately there's no consistent pattern for these types of errors right now. I suppose option 3 would be to pass the error (`ERR_get_error()`) back via an optional `unsigned long* error` parameter and also return `MaybeLocal<Value>()`.

Thoughts?

Reply:
I think this is a case where it’s just fine to throw an exception from C++. The reason that is should be avoided is that it means that code in the calling methods will have to handle and possibly propagate the exception information themselves, but that doesn’t mean that that’s a bad thing.

Reply:
Finding myself second-guessing this a bit. Most of node_crypto_common.cc returns `Undefined(env->isolate())` when an error occurs. Is the rest of that code equally as dubious?

The same goes with error handling. Almost every function I checked absorbs the OpenSSL error and returns undefined. Throwing an exception directly would require `ThrowCryptoError` and `CryptoErrorVector` to be moved over to `node_crypto_common`, which seems heavy.

Outside of status-quo, an alternate might be to return the error code in a `unsigned int* error` parameter?

Reply:
@ebickle The rule is that an empty `Maybe`/`MaybeLocal` should be returned whenever an exception is pending.

Whether it is desirable to throw an exception when encountering OpenSSL errors is another question.

> Outside of status-quo, an alternate might be to return the error code in a `unsigned int* error` parameter?

… how would that be used by the caller? If that makes the caller throw an exception, then I would strongly prefer to throw one from here and return an empty `MaybeLocal` instead.

Reply:
Moved the X509ToPEM function to node_crypto.cc since it's not used anywhere else and the ThrowCryptoError function only exists there. Updated the logic based on the recommendations here.

Reply:
> Finding myself second-guessing this a bit. Most of node_crypto_common.cc returns Undefined(env->isolate()) when an error occurs. Is the rest of that code equally as dubious?

Wouldn't quite classify it that way. node_crypto_common returns `Undefined(env->isolate())` largely because doing so preserved existing behavior when it was introduced. For many of the operations, a decision was made long ago that the actual return value for many of the openssl operations didn't matter and that returning nothing was a signal that the operation failed. In other words, failure hasn't necessarily been signaled by an explicit exception. That's definitely something that can be revisited but the goal when introducing node_crypto_common was not to change that existing behavior. It just needs to be addressed more deliberately and incrementally.

In cases where failure *is* signaled, just as when various v8 APIs return empty Maybe's, the code in node_crypto_common *does* return an empty maybe (e.g. `MaybeLocal<Value>()` with the assumption that an exception is likely pending. That assumption is not CHECK'd tho currently, as it probably should be.